libmamba/src/core/curl.cpp

```diff
@@ -307,4 +308,180 @@ namespace mamba
+        }
+        else
+        {
+            curl_multi_setopt(
+                p_handle,
+                CURLMOPT_MAX_TOTAL_CONNECTIONS,
+                static_cast<int>(max_parallel_downloads)
+            );
+        }
+    }
+
+    CURLMultiHandle::~CURLMultiHandle()
+    {
+        curl_multi_cleanup(p_handle);
+        p_handle = nullptr;
+    }
+
+
+    CURLMultiHandle::CURLMultiHandle(CURLMultiHandle&& rhs)
+        : p_handle(rhs.p_handle)
+        , m_max_parallel_downloads(rhs.m_max_parallel_downloads)
+    {
+        rhs.p_handle = nullptr;
+        rhs.m_max_parallel_downloads = 0u;
+    }
+
+    CURLMultiHandle& CURLMultiHandle::operator=(CURLMultiHandle&& rhs)
+    {
+        std::swap(p_handle, rhs.p_handle);
+        std::swap(m_max_parallel_downloads, rhs.m_max_parallel_downloads);
+        return *this;
+    }
+
+    void CURLMultiHandle::add_handle(const CURLHandle& h)
+    {
+        CURLMcode code = curl_multi_add_handle(p_handle, unwrap(h));
+        if (code != CURLM_CALL_MULTI_PERFORM)
+        {
+            if (code != CURLM_OK)
+            {
+                throw std::runtime_error(curl_multi_strerror(code));
+            }
+        }
+    }
+
+    void CURLMultiHandle::remove_handle(const CURLHandle& h)
+    {
+        curl_multi_remove_handle(p_handle, unwrap(h));
+    }
+
+    std::size_t CURLMultiHandle::perform()
+    {
+        int still_running;
+        CURLMcode code = curl_multi_perform(p_handle, &still_running);
+        if (code != CURLM_OK)
+        {
+            throw std::runtime_error(curl_multi_strerror(code));
+        }
+        return static_cast<std::size_t>(still_running);
+    }
+
+    CURLMultiHandle::response_type CURLMultiHandle::pop_message()
+    {
+        int msgs_in_queue;
+        CURLMsg* msg = curl_multi_info_read(p_handle, &msgs_in_queue);
+        if (msg != nullptr)
+        {
+            return CURLMultiResponse{ msg->easy_handle, msg->data.result, msg->msg == CURLMSG_DONE };
+        }
+        else
+        {
+            return std::nullopt;
+        }
+    }
+
+    void CURLMultiHandle::wait(size_t max_timeout)
+    {
+        long lmax_timeout = static_cast<long>(max_timeout);
+        long curl_timeout = -1;  // NOLINT(runtime/int)
+        CURLMcode code = curl_multi_timeout(p_handle, &curl_timeout);
+        if (code != CURLM_OK)
+        {
+            throw std::runtime_error(curl_multi_strerror(code));
+        }
+
+        if (curl_timeout < 0 || curl_timeout > lmax_timeout)
+        {
+            curl_timeout = lmax_timeout;
+        }
+
+        if (curl_timeout != 0)
+        {
+            int numfds;
+            code = curl_multi_wait(p_handle, NULL, 0, static_cast<int>(curl_timeout), &numfds);
+            if (code != CURLM_OK)
+            {
+                throw std::runtime_error(curl_multi_strerror(code));
+            }
+
+            if (!numfds)
+            {
+                std::this_thread::sleep_for(std::chrono::milliseconds(100));
```

Comment:
I dont get what this sleep is for? The duration is arbitrary?

Reply:
Oh good catch, it should be the timeout computed, I will fix this.

Reply:
I just ported the code that is currently in fetch, I think it was a copy paste from [there](https://curl.se/libcurl/c/curl_multi_wait.html)

Reply:
Yeah but this code is in a loop, while this function is not. If `CURLMultiHandle::wait` is called in a loop, shouldnt that last loop-sleep be in that loop too, and not in `CURLMultiHandle::wait`?

Reply:
Not sure to get you, the sleep is tighly coupled to the `curl_multi_wait` function since it depends on the numdfs returned. `wait` is meant to be called in a loop (otherwise you just wait for nothing), this `sleep` is just an implementation detail to distinguish timeouts from not having any filedescriptor left.